### ISSUE TITLE
Run reward_modeling.py by multi GPUs but model save error

### ISSUE THREAD

**Issue by ASY246** [huggingface/trl#939]
I run the example of reward model by

```accelerate launch --mixed_precision=bf16 --num_processes=8 examples/scripts/reward_modeling.py```

and use_peft = True.

Then I load the saved adaptor_model.bin by 

```
lora_weights = torch.load(os.path.join(args.reward_config.output_dir, "adapter_model.bin"))
```
and find all the tensor of layer lora_B is zero,  Is this right? If not, what could be the problem?

**Comment by lvwerra** [comment]
A question for @younesbelkada :)

**Comment by ASY246** [comment]
The script I use is as follows
```
# coding=utf-8
from dataclasses import dataclass, field
from typing import Optional

import tyro
from accelerate import Accelerator
from datasets import load_dataset
from peft import LoraConfig
from tqdm import tqdm
from transformers import AutoModelForSequenceClassification, AutoTokenizer, BitsAndBytesConfig

from trl import RewardConfig, RewardTrainer


tqdm.pandas()

@dataclass
class ScriptArguments:
    model_name: str = "/opt/ml/input/data/Llama-2-13b-chat-hf"
    """the model name"""
    dataset_text_field: str = "text"
    """the text field of the dataset"""
    eval_split: str = "none"
    """the dataset split to evaluate on; default to 'none' (no evaluation)"""
    trust_remote_code: bool = True
    """Enable `trust_remote_code`"""
    reward_config: RewardConfig = field(
        default_factory=lambda: RewardConfig(
            output_dir="/opt/ml/output",
            per_device_train_batch_size=4,
            num_train_epochs=1,
            gradient_accumulation_steps=1,
            gradient_checkpointing=True,
            learning_rate=1.41e-5,
            report_to="tensorboard",
            remove_unused_columns=False,
            optim="adamw_torch",
            logging_steps=500,
            evaluation_strategy="no",
            max_length=512,
            log_level="info"
        )
    )
    use_peft: bool = True
    """whether to use peft"""
    peft_config: Optional[LoraConfig] = field(
        default_factory=lambda: LoraConfig(
            r=16,
            lora_alpha=16,
            bias="none",
            task_type="SEQ_CLS",
            modules_to_save=["scores"],
        ),
    )
    load_in_8bit: bool = False
    """load the model in 8 bits precision"""
    load_in_4bit: bool = False
    """load the model in 4 bits precision"""
    train_dataset_path: str = "/opt/ml/input/data/reward/reward_safety_train.jsonl"
    test_dataset_path: str = "/opt/ml/input/data/reward/reward_safety_test.jsonl"

args = tyro.cli(ScriptArguments)
args.reward_config.evaluation_strategy = "steps" if args.eval_split != "none" else "no"

# Step 1: Load the model
if args.load_in_8bit and args.load_in_4bit:
    raise ValueError("You can't load the model in 8 bits and 4 bits at the same time")
elif args.load_in_8bit or args.load_in_4bit:
    quantization_config = BitsAndBytesConfig(load_in_8bit=args.load_in_8bit, load_in_4bit=args.load_in_4bit)
    # Copy the model to each device
    device_map = {"": Accelerator().local_process_index} 
else:
    device_map = None
    quantization_config = None

model = AutoModelForSequenceClassification.from_pretrained(
    args.model_name,
    quantization_config=quantization_config,
    device_map=device_map,
    trust_remote_code=args.trust_remote_code,
    num_labels=1,
    use_cache=False
)

# Step 2: Load the dataset and pre-process it
tokenizer = AutoTokenizer.from_pretrained(args.model_name)
tokenizer.pad_token = tokenizer.eos_token

reward_dataset = load_dataset("json", data_files={"train": args.train_dataset_path, "test": args.test_dataset_path})

# Tokenize chosen/rejected pairs of inputs
# Adapt this section to your needs for custom datasets

def preprocess_function(examples):
    new_examples = {
        "input_ids_chosen": [],
        "attention_mask_chosen": [],
        "input_ids_rejected": [],
        "attention_mask_rejected": [],
    }
    for chosen, rejected in zip(examples["chosen"], examples["rejected"]):
        tokenized_chosen = tokenizer(chosen, truncation=True) 
        tokenized_rejected = tokenizer(rejected, truncation=True)
        new_examples["input_ids_chosen"].append(tokenized_chosen["input_ids"])
        new_examples["attention_mask_chosen"].append(tokenized_chosen["attention_mask"])
        new_examples["input_ids_rejected"].append(tokenized_rejected["input_ids"])
        new_examples["attention_mask_rejected"].append(tokenized_rejected["attention_mask"])

    return new_examples

# Preprocess the dataset and filter out examples that are longer than script_args.max_length
reward_dataset = reward_dataset.map(
    preprocess_function,
    batched=True,
    num_proc=20
)
reward_dataset = reward_dataset.filter(
    lambda x: len(x["input_ids_chosen"]) <= args.reward_config.max_length and len(x["input_ids_rejected"]) <= args.reward_config.max_length
)

# Step 4: Define the LoraConfig
if args.use_peft:
    peft_config = args.peft_config
else:
    peft_config = None

# Step 5: Define the Trainer
trainer = RewardTrainer(
    model=model,
    tokenizer=tokenizer,
    args=args.reward_config,
    train_dataset=reward_dataset["train"],
    eval_dataset=reward_dataset["test"],
    peft_config=peft_config,
)

trainer.train()


import os
import torch

if Accelerator().local_process_index == 0:
    for name, param in trainer.model.named_parameters():
        if param.requires_grad:
            print("before save trained name:{}, param:{}".format(name, param))
    lora_weights = torch.load(os.path.join(args.reward_config.output_dir, "checkpoint-3000", "adapter_model.bin"))
    for k, v in lora_weights.items():
        print("lora layer key:{}, value:{}, sum:{}".format(k, v, v.sum()))
        l, r = v.shape
        tmp=torch.zeros(l, r).to(Accelerator().local_process_index)
        print("is equal:{}".format(torch.equal(tmp, v)))
```

**Comment by ASY246** [comment]
For details，I write a callback class to print the model named_parameter after training as follows:
```
class CustomPeftSavingCallback(TrainerCallback):
    def on_save(self, args, state, control, **kwargs):
        if args.should_save:
            for name, param in kwargs["model"].named_parameters():
                if param.requires_grad:
                    print("callback before save trained name:{}, param:{}".format(name, param))
callbacks = [CustomPeftSavingCallback()]

trainer = RewardTrainer(
    model=model,
    tokenizer=tokenizer,
    args=args.reward_config,
    train_dataset=reward_dataset["train"],
    eval_dataset=reward_dataset["test"],
    peft_config=peft_config,
    callbacks=callbacks
)
```
And I still find the elements in Lora_B are zeros like this

```
[2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.q_proj.lora_A.default.weight, param:Parameter containing:
[2023-11-02 16:02:27] tensor([[ 0.0012,  0.0120, -0.0004,  ...,  0.0083,  0.0048,  0.0112],
[2023-11-02 16:02:27]         [-0.0037, -0.0091, -0.0100,  ...,  0.0121,  0.0060,  0.0136],
[2023-11-02 16:02:27]         [-0.0038, -0.0003, -0.0087,  ..., -0.0011,  0.0058, -0.0066],
[2023-11-02 16:02:27]         ...,
[2023-11-02 16:02:27]         [-0.0094, -0.0040, -0.0122,  ...,  0.0026, -0.0064, -0.0102],
[2023-11-02 16:02:27]         [ 0.0002,  0.0007, -0.0045,  ...,  0.0100,  0.0073,  0.0044],
[2023-11-02 16:02:27]         [-0.0129, -0.0044, -0.0058,  ..., -0.0033,  0.0063,  0.0016]],
[2023-11-02 16:02:27]        device='cuda:0', requires_grad=True)
[2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.q_proj.lora_B.default.weight, param:Parameter containing:
[2023-11-02 16:02:27] tensor([[0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         ...,
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0', requires_grad=True)
[2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.v_proj.lora_A.default.weight, param:Parameter containing:
[2023-11-02 16:02:27] tensor([[ 0.0070,  0.0108, -0.0006,  ..., -0.0039, -0.0066, -0.0138],
[2023-11-02 16:02:27]         [-0.0048, -0.0066, -0.0122,  ..., -0.0134,  0.0062, -0.0112],
[2023-11-02 16:02:27]         [-0.0059,  0.0069, -0.0076,  ...,  0.0123,  0.0122,  0.0048],
[2023-11-02 16:02:27]         ...,
[2023-11-02 16:02:27]         [-0.0107, -0.0129,  0.0038,  ..., -0.0087,  0.0112, -0.0007],
[2023-11-02 16:02:27]         [-0.0076,  0.0088, -0.0077,  ...,  0.0024, -0.0116,  0.0097],
[2023-11-02 16:02:27]         [-0.0017,  0.0056,  0.0014,  ...,  0.0096, -0.0005, -0.0127]],
[2023-11-02 16:02:27]        device='cuda:0', requires_grad=True)
[2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.v_proj.lora_B.default.weight, param:Parameter containing:
[2023-11-02 16:02:27] tensor([[0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         ...,
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
[2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0', requires_grad=True)
```

But I am sure the loss during training is decreasing, even though I remove the modules_to_save=["scores"]. Does anyone could explain the reasons?

**Comment by sworddish** [comment]
> For details，I write a callback class to print the model named_parameter after training as follows:
> 
> ```
> class CustomPeftSavingCallback(TrainerCallback):
>     def on_save(self, args, state, control, **kwargs):
>         if args.should_save:
>             for name, param in kwargs["model"].named_parameters():
>                 if param.requires_grad:
>                     print("callback before save trained name:{}, param:{}".format(name, param))
> callbacks = [CustomPeftSavingCallback()]
> 
> trainer = RewardTrainer(
>     model=model,
>     tokenizer=tokenizer,
>     args=args.reward_config,
>     train_dataset=reward_dataset["train"],
>     eval_dataset=reward_dataset["test"],
>     peft_config=peft_config,
>     callbacks=callbacks
> )
> ```
> 
> And I still find the elements in Lora_B are zeros like this
> 
> ```
> [2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.q_proj.lora_A.default.weight, param:Parameter containing:
> [2023-11-02 16:02:27] tensor([[ 0.0012,  0.0120, -0.0004,  ...,  0.0083,  0.0048,  0.0112],
> [2023-11-02 16:02:27]         [-0.0037, -0.0091, -0.0100,  ...,  0.0121,  0.0060,  0.0136],
> [2023-11-02 16:02:27]         [-0.0038, -0.0003, -0.0087,  ..., -0.0011,  0.0058, -0.0066],
> [2023-11-02 16:02:27]         ...,
> [2023-11-02 16:02:27]         [-0.0094, -0.0040, -0.0122,  ...,  0.0026, -0.0064, -0.0102],
> [2023-11-02 16:02:27]         [ 0.0002,  0.0007, -0.0045,  ...,  0.0100,  0.0073,  0.0044],
> [2023-11-02 16:02:27]         [-0.0129, -0.0044, -0.0058,  ..., -0.0033,  0.0063,  0.0016]],
> [2023-11-02 16:02:27]        device='cuda:0', requires_grad=True)
> [2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.q_proj.lora_B.default.weight, param:Parameter containing:
> [2023-11-02 16:02:27] tensor([[0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         ...,
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0', requires_grad=True)
> [2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.v_proj.lora_A.default.weight, param:Parameter containing:
> [2023-11-02 16:02:27] tensor([[ 0.0070,  0.0108, -0.0006,  ..., -0.0039, -0.0066, -0.0138],
> [2023-11-02 16:02:27]         [-0.0048, -0.0066, -0.0122,  ..., -0.0134,  0.0062, -0.0112],
> [2023-11-02 16:02:27]         [-0.0059,  0.0069, -0.0076,  ...,  0.0123,  0.0122,  0.0048],
> [2023-11-02 16:02:27]         ...,
> [2023-11-02 16:02:27]         [-0.0107, -0.0129,  0.0038,  ..., -0.0087,  0.0112, -0.0007],
> [2023-11-02 16:02:27]         [-0.0076,  0.0088, -0.0077,  ...,  0.0024, -0.0116,  0.0097],
> [2023-11-02 16:02:27]         [-0.0017,  0.0056,  0.0014,  ...,  0.0096, -0.0005, -0.0127]],
> [2023-11-02 16:02:27]        device='cuda:0', requires_grad=True)
> [2023-11-02 16:02:27] callback before save trained name:base_model.model.model.layers.0.self_attn.v_proj.lora_B.default.weight, param:Parameter containing:
> [2023-11-02 16:02:27] tensor([[0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         ...,
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.],
> [2023-11-02 16:02:27]         [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:0', requires_grad=True)
> ```
> 
> But I am sure the loss during training is decreasing, even though I remove the modules_to_save=["scores"]. Does anyone could explain the reasons?

I met the same issue using dpo_trainer with zero stage-3,  loss was decreasing, but the model saved have lots of empty parameters(torch.Size([0])  throughout multiple layers. 

I thought it could because of some bugs during model saving period since zero stage 3 need to collect the sharding paras after training process finished, still could not figure it out at the moment.

**Comment by Elfsong** [comment]
#742

**Comment by Elfsong** [comment]
@ASY246 Given that PR #956 has been merged, could you try out if the problem has been solved as well?

**Comment by ASY246** [comment]
> @ASY246 Given that PR #956 has been merged, could you try out if the problem has been solved as well?

My problem still exists. I wonder if it's because actually I'm running the reward model training script, and this change only adjusts the DPO part without modifying the logic of the reward model.
